core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
@vanzin pinged me offline about having an RPC happening in the listener thread for each executor being reported idle that can be a huge bottle neck. I am planning to make all code in this thread asynchronous, and make the maps and sets used in this thread synchronized, to avoid blocking the listener thread. 

Thoughts - @andrewor14, @vanzin ?